### PR TITLE
fix: Add additional required expression for natural join

### LINKED ISSUE #11635: A valid SQL query returned 'Schema error: ...' (SQLancer)
### Describe the bug

The following SQL query should be valid, not it's returning an error

Run in datafusion-cli:
```
/*DML*/CREATE TABLE t0(v0 BIGINT, v1 DOUBLE);
/*DML*/CREATE TABLE t1(v0 BIGINT, v1 STRING);
/*DML*/CREATE TABLE t5(v0 BIGINT, v1 STRING, v2 BIGINT, v3 STRING, v4 BOOLEAN);
```
```
> SELECT *
FROM t1
NATURAL JOIN t5
INNER JOIN t0 ON (t0.v1 + t5.v0) > 0
WHERE t0.v1 = t1.v0;

Schema error: No field named t5.v0. Valid fields are t1.v0, t1.v1, t5.v2, t5.v3, t5.v4, "CAST(t1.v0 AS Float64)", t0.v0, t0.v1.
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

Found by SQLancer https://github.com/apache/datafusion/issues/11030

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #11635

### CODE REVIEW

File: datafusion/expr/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -1532,13 +1531,28 @@ pub fn wrap_projection_for_join_if_necessary(
     let need_project = join_keys.iter().any(|key| !matches!(key, Expr::Column(_)));
     let plan = if need_project {
         let mut projection = expand_wildcard(input_schema, &input, None)?;
+        // #11635 when we are dealing with join, and the children schema has diffrencens with join schema,
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/builder.rs:1534]:
```suggestion
        // #11635 when we are dealing with join, and the children schema has differences with join schema,
```

File: datafusion/expr/src/logical_plan/builder.rs [hunk 2]
```diff
@@ -1507,6 +1505,7 @@ fn table_source(table_schema: &Schema) -> Arc<dyn TableSource> {
 pub fn wrap_projection_for_join_if_necessary(
     join_keys: &[Expr],
     input: LogicalPlan,
+    upper_exprs: &[&Expr],
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/builder.rs:1508]: Could you please add some documentation about what this parameter is being used for? 

For example, it seems like it represents the expressions required by the parent join. And the change is that this function ensures that all expressions in `upper_exprs` are included in the output `LogicalPlan`?

**Review by alamb — APPROVED** [review]
Thanks @Lordworms  -- this code looks reasonable to me. I had some suggestions / requests for some more comments.

Maybe @jonahgao  has ideas of how to improve this too

File: datafusion/expr/src/logical_plan/builder.rs (hunk 2 above)
Comment (Lordworms) [reply datafusion/expr/src/logical_plan/builder.rs:1508]: sure, I'll add that

**Comment by jonahgao** [comment]
I checked that the final logical plan is correct, but the calling of [expand_wildcard](https://github.com/apache/datafusion/blob/d010ce90f40f2866904a4eea563afbbff72497cc/datafusion/expr/src/logical_plan/builder.rs#L1534) in `wrap_projection_for_join_if_necessary` breaks it.  `expand_wildcard` removes duplicate columns in the using join, but this column is needed by the parent plan.

The use of `expand_wildcard` was introduced in https://github.com/apache/datafusion/pull/4443/. I speculate that it was a refactor of the following old implementation.
```rust
let mut projection = vec![Expr::Wildcard];
        projection.extend(expr_join_keys.into_iter());
        LogicalPlanBuilder::from(input)
            .project(projection)?
            .build()?
```
Its purpose should be to re-project all the outputs of the input plan.

So I think another simpler fix might be to replace `expand_wildcard` with [this](https://github.com/jonahgao/datafusion/commit/591f24680a8ef83d18da990baf2d38a51c288f6a).

**Comment by Lordworms** [comment]
> I checked that the final logical plan is correct, but the calling of [expand_wildcard](https://github.com/apache/datafusion/blob/d010ce90f40f2866904a4eea563afbbff72497cc/datafusion/expr/src/logical_plan/builder.rs#L1534) in `wrap_projection_for_join_if_necessary` breaks it. `expand_wildcard` removes duplicate columns in the using join, but this column is needed by the parent plan.
> 
> The use of `expand_wildcard` was introduced in #4443. I speculate that it was a refactor of the following old implementation.
> 
> ```rust
> let mut projection = vec![Expr::Wildcard];
>         projection.extend(expr_join_keys.into_iter());
>         LogicalPlanBuilder::from(input)
>             .project(projection)?
>             .build()?
> ```
> 
> Its purpose should be to re-project all the outputs of the input plan.
> 
> So I think another simpler fix might be to replace `expand_wildcard` with [this](https://github.com/jonahgao/datafusion/commit/591f24680a8ef83d18da990baf2d38a51c288f6a).

Yes, I am aware of 

> expand_wildcard

I agree, It's is much simpler, I was thinking about how to add missing expressions that  are excluded by `expand_wildcard` but haven't thought about replace it directly. Thanks for this.

**Comment by Lordworms** [comment]
> > I checked that the final logical plan is correct, but the calling of [expand_wildcard](https://github.com/apache/datafusion/blob/d010ce90f40f2866904a4eea563afbbff72497cc/datafusion/expr/src/logical_plan/builder.rs#L1534) in `wrap_projection_for_join_if_necessary` breaks it. `expand_wildcard` removes duplicate columns in the using join, but this column is needed by the parent plan.
> > The use of `expand_wildcard` was introduced in #4443. I speculate that it was a refactor of the following old implementation.
> > ```rust
> > let mut projection = vec![Expr::Wildcard];
> >         projection.extend(expr_join_keys.into_iter());
> >         LogicalPlanBuilder::from(input)
> >             .project(projection)?
> >             .build()?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Its purpose should be to re-project all the outputs of the input plan.
> > So I think another simpler fix might be to replace `expand_wildcard` with [this](https://github.com/jonahgao/datafusion/commit/591f24680a8ef83d18da990baf2d38a51c288f6a).
> 
> 
> > expand_wildcard
> 
> I agree, It's is much simpler, I was thinking about how to add missing expressions that are excluded by `expand_wildcard` but haven't thought about replace it directly. Thanks for this.

**Review by jonahgao — APPROVED** [review]
Looks good to me!

File: datafusion/expr/src/logical_plan/builder.rs [hunk 3]
```diff
@@ -1531,7 +1531,12 @@ pub fn wrap_projection_for_join_if_necessary(
 
     let need_project = join_keys.iter().any(|key| !matches!(key, Expr::Column(_)));
     let plan = if need_project {
-        let mut projection = expand_wildcard(input_schema, &input, None)?;
+        // Include all columns from the input and extend them with the join keys
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/builder.rs:1534]: 😍  that is beautiful.

**Review by alamb — APPROVED** [review]
Thank you @Lordworms and @jonahgao  -- this is a beautifully simple PR now. 🏆

**Pull request merged**